### PR TITLE
Don't count recursive function calls into function trace depth

### PR DESCRIPTION
**Pull request by piotmag769**
Closes https://github.com/foundry-rs/starknet-foundry/issues/2133

## Introduced changes

- During trace building don't count consecutive calls of functions with the same name into function trace depth. Instead count how many additional times a function were called and skip that many return statements.

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation (README.md)
- [ ] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cairo-profiler/src/trace_reader/function_trace_builder.rs [hunk 1]
```diff
@@ -113,7 +114,8 @@ pub fn collect_profiling_info(
             }
         };
 
-        let user_function_idx = user_function_idx(sierra_statement_idx, program);
+        let user_function_name = user_function_name(sierra_statement_idx, program)
+            .to_displayable_function_name(function_level_config.split_generics);
```
Comment (piotmag769) [inline crates/cairo-profiler/src/trace_reader/function_trace_builder.rs:118]: Passing the `split_generics` here may seem weird but it actually makes sense. If a users choses not to split them then e.g. `func<felt252>` and `func<u8>` look the same to them, so if they are called consecutively one after the other they should be aggregated for the purpose of function stack depth - they see them as a single function on the stack anyways.

File: crates/cairo-profiler/src/trace_reader/function_trace_builder.rs [hunk 2]
```diff
@@ -126,8 +129,19 @@ pub fn collect_profiling_info(
                     Ok(CoreConcreteLibfunc::FunctionCall(_))
                 ) {
                     // Push to the stack.
+                    if let Some(stack_element) = function_stack.last_mut() {
+                        if user_function_name == stack_element.caller_function_name {
+                            stack_element.recursive_calls_count += 1;
+                            continue;
+                        }
+                    }
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-profiler/src/trace_reader/function_trace_builder.rs:137]: Please abstract it as stack operation `push`. Please do the same for `pop`. Recursive calls counting should be an implementation detail of the stack

File: crates/cairo-profiler/src/trace_reader/function_trace_builder.rs (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply crates/cairo-profiler/src/trace_reader/function_trace_builder.rs:119]: I generally think the problem here is that you pass those dependencies around. You should have a struct that holds program and split_generics, some kind of database, and perform a query with `sierra_statement_idx` on that.

**Review by MaksymilianDemitraszek — APPROVED** [review]
Please address the comments

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
.

File: crates/cairo-profiler/src/trace_reader/function_stack_trace.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    // this remains reliable while `function_stack` does not.
+    real_function_stack_depth: usize,
+    // Constant through existence of the object.
+    max_function_trace_depth: usize,
+}
+
+pub enum PoppedElement {
+    RegularFunction(StackElement),
+    HiddenOrRecursiveFunction,
+}
```
Comment (MaksymilianDemitraszek) [inline crates/cairo-profiler/src/trace_reader/function_stack_trace.rs:23]:
```suggestion
pub enum Element {
    RegularFunction(StackElement),
    HiddenFunction,
}
```

**Review by MaksymilianDemitraszek — APPROVED** [review]

**Pull request merged**